File: solutions/haskell/029-valid-sudoku.hs
{-
  Problem 29: Valid Sudoku (LeetCode 36)
  Difficulty: Med
  Language: Haskell
-}
import qualified Data.Set as Set

isValidSudoku :: [[Char]] -> Bool
isValidSudoku board =
  let entries = [ (i, j, c)
                | (i, row) <- zip [0..] board
                , (j, c)   <- zip [0..] row
                , c /= '.' ]
      keys (i, j, c) =
        [ "r" ++ show i ++ ":" ++ [c]
        , "c" ++ show j ++ ":" ++ [c]
        , "b" ++ show (i `div` 3) ++ "," ++ show (j `div` 3) ++ ":" ++ [c]
        ]
      allKeys = concatMap keys entries
  in length allKeys == Set.size (Set.fromList allKeys)

File: solutions/elixir/029-valid-sudoku.ex
# Problem 29: Valid Sudoku (LeetCode 36)
# Difficulty: Med
# Language: Elixir
# 
defmodule ValidSudoku do
  def is_valid?(board) do
    entries =
      for {row, i} <- Enum.with_index(board),
          {cell, j} <- Enum.with_index(row),
          cell != ".",
          do: {i, j, cell}

    keys =
      Enum.flat_map(entries, fn {i, j, c} ->
        ["r#{i}:#{c}", "c#{j}:#{c}",
         "b#{div(i,3)},#{div(j,3)}:#{c}"]
      end)

    length(keys) == MapSet.size(MapSet.new(keys))
  end
end
